### PR TITLE
Support KILT for Pyserini's h/d/search

### PR DESCRIPTION
**Pull request by yuxuan-ji**
Changes to the pyserini h/d/search scripts:
- Adds non-required flag `--topics-format` which defaults to `default`. Other option is `kilt`
- Adds non-required flag `--tokenizer` to dsearch. This is useful when the model used does not have a tokenizer specified, such as KILT's (it uses the bert-base-uncased tokenizer).
- Adds non-required flag `--output-format` which defaults to `trec`. Other options are `msmarco` and `kilt`.
- Deprecated the `--msmarco` flag, as it is now specified through the above.

Adds the KILT evaluation script to `pyserini.eval`

Introduces the following abstractions:
- There is now a `QueryIterator` class for each topic format. It does the job of doing any pre/post processing required when loading/iterating through the dataset. Notably, KILT does some post-processing on the queries, and needs to be loaded from a file currently.
- There is now a `OutputWriter` class for TREC, MSMARCO, and KILT's formats.

Added integration tests are in: 
https://github.com/yuxuan-ji/pyserini/pull/1

Successfully ran integration tests for (can run more if needed, they take quite a while though):
- DPR curated, nq (hsearch + dsearch)
- TCT-Colbert, msmarco doc (hsearch)

### CODE REVIEW

File: pyserini/query_iterator.py [hunk 1]
```diff
@@ -1,13 +1,82 @@
...
+        return cls(topics_path)
+
+
+def get_query_iterator(topics_path: str, query_format: TopicsFormat):
+    mapping = {
+        TopicsFormat.DEFAULT: DefaultQueryIterator,
+        TopicsFormat.KILT_DRQA: KiltQueryIterator,
+        TopicsFormat.KILT_DPR: KiltQueryIterator
+    }
+    return mapping[query_format].from_topics(topics_path)
```
Comment (yuxuan-ji) [inline pyserini/query_iterator.py:82]: diff too big

File: pyserini/output_writer.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        provenance = []
+        for docid, rank, score, _ in self.hits_iterator(hits):
+            provenance.append({"wikipedia_id": docid})
+        datapoint["output"] = [{"provenance": provenance}]
+        json.dump(datapoint, self.file)
+        self.file.write('\n')
+
+
+class KiltDprWriter(OutputWriter):
+    pass
```
Comment (yuxuan-ji) [inline pyserini/output_writer.py:85]: TODO

**Comment by lintool** [comment]
hey @ronakice can you coordinate with @yuxuan-ji to see where this is going? should be be part of next release? https://github.com/castorini/pyserini/projects/1

**Comment by lintool** [comment]
We'll probably need integration tests, like https://github.com/castorini/pyserini/tree/master/integrations

To make sure things don't break moving forward...

File: pyserini/eval/evaluate_kilt_retrieval.py [hunk 3]
```diff
@@ -0,0 +1,366 @@
+# NOTE: This code is taken from the original KILT library's retrieval evaluation script
+# https://github.com/facebookresearch/KILT/blob/9bcb119a7ed5fda88826058b062d0e45c726c676/kilt/eval_retrieval.py
```
Comment (yuxuan-ji) [inline pyserini/eval/evaluate_kilt_retrieval.py:2]: eval script taken from KILT

File: pyserini/search/_base.py [hunk 4]
```diff
@@ -142,16 +142,6 @@ def get_topics(collection_name):
...
-    elif os.path.exists(collection_name):
-        t = {}
-        for line in open(collection_name).readlines():
-            topic, text = line.rstrip().split('\t')
-            try:
-                topic = int(topic)
-            except ValueError:
-                pass
-            t[topic]['title'] = text
-        return t
```
Comment (yuxuan-ji) [inline pyserini/search/_base.py:154]: moving this to the `QueryIterator` class

File: pyserini/external_query_info.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        "total_queries": 2837
+    },
+    "hotpotqa-dev-kilt": {
+        "description": "KILT HotpotQA queries",
+        "urls": ["http://dl.fbaipublicfiles.com/KILT/hotpotqa-dev-kilt.jsonl"],
+        "md5": "3ebc5eeaa5572ec29451eb4b66c29333",
+        "size (bytes)": 3971321,
+        "total_queries": 5600
+    },
+    "triviaqa-dev-kilt": {},
```
Comment (yuxuan-ji) [inline pyserini/external_query_info.py:58]: to be added

File: pyserini/search/_base.py (hunk 4 above)
Comment (MXueguang) [reply pyserini/search/_base.py:154]: ah yeah your are right

File: pyserini/query_iterator.py [hunk 6]
```diff
@@ -1,23 +1,152 @@
...
+from urllib.error import HTTPError, URLError
+
+
+@unique
+class TopicsFormat(Enum):
+    DEFAULT = 'default'
+    KILT = 'kilt'
+
+
+class QueryIterator(ABC):
```
Comment (yuxuan-ji) [inline pyserini/query_iterator.py:33]: diff too big, but here's the queryiterator class

File: scripts/kilt/anserini_retriever.py [hunk 7]
```diff
@@ -63,7 +63,7 @@ def _get_predictions_thread(arguments):
         doc_scores = []
 
         if use_bigrams:
-            tokens = filter(lambda word: word not in STOPWORDS, word_tokenize(query))
+            tokens = filter(lambda word: word.lower() not in STOPWORDS, word_tokenize(query))
```
Comment (yuxuan-ji) [inline scripts/kilt/anserini_retriever.py:66]: fix to previous script

File: scripts/kilt/convert_kilt_dpr_to_pyserini_format.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+
+# All files required for this script can be found at:
+# https://github.com/facebookresearch/KILT/tree/master/kilt/retrievers#download-models-1
+# Note: Use this script
+# https://github.com/huggingface/transformers/blob/053efc5d2d2e87833e9b7290a0dd83fa77cd6ae8/src/transformers/models/dpr/convert_dpr_original_checkpoint_to_pytorch.py
+# to convert KILT's dpr_multi_set_f_bert.0 model into a PyTorch checkpoint
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Convert KILT-dpr corpus into the index & docid file read by pyserini')
```
Comment (yuxuan-ji) [inline scripts/kilt/convert_kilt_dpr_to_pyserini_format.py:18]: script to encode kilt topics using DPR

File: scripts/kilt/convert_kilt_to_document_jsonl.py [hunk 9]
```diff
@@ -26,7 +26,7 @@
             doc["id"] = raw["_id"]
             doc["contents"] = "".join(raw["text"])
             if args.bigrams:
-                tokens = filter(lambda word: word not in STOPWORDS, word_tokenize(doc["contents"]))
+                tokens = filter(lambda word: word.lower() not in STOPWORDS, word_tokenize(doc["contents"]))
```
Comment (yuxuan-ji) [inline scripts/kilt/convert_kilt_to_document_jsonl.py:29]: fix to previous script

File: scripts/kilt/convert_kilt_to_passage_jsonl.py [hunk 10]
```diff
@@ -42,7 +42,7 @@
                 doc["id"] = f"{raw['_id']}-{i}"
                 p = texts[i]
                 if args.bigrams:
-                    tokens = filter(lambda word: word not in STOPWORDS, word_tokenize(p))
+                    tokens = filter(lambda word: word.lower() not in STOPWORDS, word_tokenize(p))
```
Comment (yuxuan-ji) [inline scripts/kilt/convert_kilt_to_passage_jsonl.py:45]: fix to previous script

File: scripts/kilt/encode_kilt_topics.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
+import argparse
+import pandas as pd
+
+from tqdm import tqdm
+from pyserini.query_iterator import get_query_iterator, TopicsFormat
+from pyserini.dsearch import DprQueryEncoder
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Compute embeddings for KILT topics')
```
Comment (yuxuan-ji) [inline scripts/kilt/encode_kilt_topics.py:10]: script to encoded KILT topics w/ dpr

**Review by TambourineMan42 — APPROVED** [review]
Okay, most of this looks good, didn't see anything glaringly wrong, @MXueguang do you have any suggestions?

**Review by ronakice — APPROVED** [review]
Okay switched back to my other system, a yes from the right account! An LGTM from the right account! 😄

**Comment by MXueguang** [comment]
lgtm. but we'd better run the entire integration tests on tuna/orca before we merge it

**Comment by lintool** [comment]
Agreed. @yuxuan-ji can you fix conflicts? I'll run tests, and then I can merge.

**Comment by MXueguang** [comment]
just fyi @lintool, the integration tests for this PR is here https://github.com/yuxuan-ji/pyserini/pull/1
@yuxuan-ji since we review the above changes, maybe merge the update for tests into this branch too?
(then we can run test on the HEAD of this branch)

**Comment by lintool** [comment]
Sure!

**Comment by lintool** [comment]
I'm currently verifying #513 right now. Will queue this up next.

**Comment by yuxuan-ji** [comment]
@MXueguang @lintool merged the integration tests into this PR, so tests can be ran - as a result it does make this PR quite bloated unfortunately 

one thing: I have an anserini index that needs to be pushed to https://git.uwaterloo.ca/jimmylin/anserini-indexes for the kilt test to work, it's currently sitting on tuna, what would be the best way to upload it?

**Comment by lintool** [comment]
Hi all, I've started running integration tests on my iMac Pro. Please no more pushes to this branch. Will report back when done.

**Comment by lintool** [comment]
Regressions just completed:

```
======================================================================
ERROR: test_kilt_search (test_kilt.TestSearchIntegration)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jimmylin/workspace/pyserini/integrations/test_kilt.py", line 63, in test_kilt_search
    self.assertAlmostEqual(score, 0.3821, delta=0.0001)
  File "/anaconda3/envs/python36/lib/python3.6/unittest/case.py", line 861, in assertAlmostEqual
    if abs(first - second) <= delta:
TypeError: unsupported operand type(s) for -: 'NoneType' and 'float'

----------------------------------------------------------------------
Ran 104 tests in 47456.780s

FAILED (errors=1)
```

The failure is due to the fact that the Anserini index hasn't been installed yet. I will merge, then circle back to fix.

**Pull request merged**